firmware/applications/SignBuddy/imu.c

```diff
@@ -0,0 +1,119 @@
+#include "imu.h"
+
+#include "board.h"
+#include "common.h"
+#include "i2c.h"
+#include "logger.h"
+#include "bno055.h"
+
+#define PROCESS_PERIOD_MS    1000
+
+typedef struct {
+  uint32_t                last_ticks;
+  i2c_t                   i2c_instance;
+  struct   bno055_t       bno055;
+  struct   bno055_accel_t bno055_accel_xyz;
+  struct   bno055_mag_t   bno055_mag_xyz;
+  struct   bno055_gyro_t  bno055_gyro_xyz;
+} state_t;
+
+static state_t state;
+
+static void hw_init(void)
+{
+  IMU_I2C_CLK_EN();
+  IMU_I2C_GPIO_CLK_EN();
+
+  LL_GPIO_InitTypeDef gpio_config = { 0 };
+  gpio_config.Pin = IMU_I2C_SCL_PIN | IMU_I2C_SDA_PIN;
+  gpio_config.Mode = LL_GPIO_MODE_ALTERNATE;
+  gpio_config.Speed = LL_GPIO_SPEED_FREQ_VERY_HIGH;
+  gpio_config.OutputType = LL_GPIO_OUTPUT_OPENDRAIN;
+  gpio_config.Pull = LL_GPIO_PULL_NO;
+  gpio_config.Alternate = IMU_I2C_GPIO_AF;
+  LL_GPIO_Init(IMU_I2C_GPIO_PORT, &gpio_config);
+
+  LL_I2C_InitTypeDef i2c_config = { 0 };
+  LL_I2C_EnableAutoEndMode(IMU_I2C);
+  LL_I2C_DisableOwnAddress2(IMU_I2C);
+  LL_I2C_DisableGeneralCall(IMU_I2C);
+  LL_I2C_EnableClockStretching(IMU_I2C);
+  i2c_config.PeripheralMode = LL_I2C_MODE_I2C;
+  i2c_config.Timing = 0x00300F38;
+  i2c_config.AnalogFilter = LL_I2C_ANALOGFILTER_ENABLE;
+  i2c_config.DigitalFilter = 0;
+  i2c_config.OwnAddress1 = 0;
+  i2c_config.TypeAcknowledge = LL_I2C_ACK;
+  i2c_config.OwnAddrSize = LL_I2C_OWNADDRESS1_7BIT;
+  LL_I2C_SetOwnAddress2(IMU_I2C, 0, LL_I2C_OWNADDRESS2_NOMASK);
+
+  i2c_init(&state.i2c_instance, IMU_I2C, &i2c_config);
+}
+
+static inline int8_t write(uint8_t slave_addr, uint8_t reg_addr, uint8_t *data, uint8_t length)
+{
+  i2c_write(&state.i2c_instance, slave_addr << 1, reg_addr, data, (uint16_t) length);
+  return 0;
+}
+
+static inline int8_t read(uint8_t slave_addr, uint8_t reg_addr, uint8_t *data, uint8_t length)
+{
+  i2c_read(&state.i2c_instance, slave_addr << 1, reg_addr, data, (uint16_t) length);
+  return 0;
+}
```

Comment:
i think we should rename these since read/write are both standard c functions already

Reply:
okay sure